File: SettingsManager/mainframe.cpp
#include "mainframe.h"

#include "main.h"
#include "dlgoptions.h"
#include "settingsmanager.h"

#include <wx/display.h>

#ifndef __WXMSW__
#include "wxwin32x32.xpm"
#endif // __WXMSW__

MainFrame::MainFrame(const wxString& title) : wxFrame(NULL, -1, title),
	m_settings(SettingsManager::Get())
{
#ifdef __WXDEBUG__
	wxPrintf(_T("Creating a \"MainFrame\" object\n"));
#endif // __WXDEBUG__

	SetIcon(wxICON(appIcon)); // Defining app icon

	CreateControls();

	int iStartPos = m_settings.GetMainWndStartupPos();
	wxSize szStartSize = m_settings.GetMainWndStartupSize();
	wxPoint ptStartPos;
	m_settings.GetMainWndStartupPos(ptStartPos);
	if (iStartPos==wxALIGN_NOT)
	{
		if (szStartSize == wxDefaultSize)
		{
			if (ptStartPos == wxDefaultPosition)
			{
				Maximize(true);
			}
			else
			{
				SetSize(MAINFRAME_MIN_SIZE);
				CenterOnScreen();
			}
		}
		else
		{
			Move(ptStartPos);
			SetSize(szStartSize);
		}
	}
	else
	{

		wxDisplay d;
		wxRect rcD=d.GetClientArea();
		int iWScr=rcD.GetWidth();
		int iHScr=rcD.GetHeight();
		wxPoint pt=wxDefaultPosition;
		if (szStartSize == wxDefaultSize)
			szStartSize = MAINFRAME_MIN_SIZE;
		pt.x=(((iStartPos&wxLEFT)==wxLEFT)?0:((iStartPos&wxRIGHT)==wxRIGHT)?iWScr-szStartSize.GetWidth():(iWScr-szStartSize.GetWidth())/2);
		pt.y=(((iStartPos&wxTOP)==wxTOP)?0:((iStartPos&wxBOTTOM)==wxBOTTOM)?iHScr-szStartSize.GetHeight():(iHScr-szStartSize.GetHeight())/2);
		Move(pt);
		SetSize(szStartSize);
	}
	SetMinSize(MAINFRAME_MIN_SIZE);

	ConnectControls();
}

MainFrame::~MainFrame()
{
#ifdef __WXDEBUG__
	wxPrintf(_T("Destroying a \"MainFrame\" object\n"));
#endif // __WXDEBUG__
}

void MainFrame::CreateControls()
{
	// Status bar
	CreateStatusBar(1);
	wxString sTxt=wxGetApp().GetBuildInfos(true);
	SetStatusText(sTxt);

	// Add a panel to the frame (needed under Windows to have a nice frame)
	wxPanel *pnl=new wxPanel(this, -1);

	wxBoxSizer *szrMain=new wxBoxSizer(wxVERTICAL);
		m_btnPrefs=new wxButton(pnl, -1, _T("Edit Settings"));
		szrMain->Add(m_btnPrefs, 0, wxALL, 5);
	pnl->SetSizer(szrMain);
}

void MainFrame::ConnectControls()
{
	// General events handlers
	Bind(wxEVT_SIZE, &MainFrame::OnSize, this);
	Bind(wxEVT_MOVE, &MainFrame::OnMove, this);

	m_btnPrefs->Bind(wxEVT_BUTTON, &MainFrame::OnBtnPrefsClicked, this);
}

void MainFrame::OnSize(wxSizeEvent& event)
{
	if (!IsShown()) return;
	if (IsMaximized())
	{
		m_settings.SetLastWindowRect(wxDefaultPosition, wxDefaultSize);
	}
	else
	{
		m_settings.SetLastWindowRect(GetPosition(), GetSize());
	}
	event.Skip();
}

void MainFrame::OnMove(wxMoveEvent& event)
{
	if (!IsShown()) return;
	m_settings.SetLastWindowRect(GetPosition(), GetSize());
	event.Skip();
}

void MainFrame::OnBtnPrefsClicked(wxCommandEvent &event)
{
	DlgOptions dlg(this);
	dlg.ShowModal();
}

File: SettingsManager/Python/MainFrame.py
import wx
from SettingsManager import SettingsManager
from DlgSettings import DlgSettings

class MainFrame(wx.Frame):

    def __init__(self):
        wx.Frame.__init__(self, None, title=wx.GetApp().GetAppName())
        self.SetIcon(wx.Icon('../../common/wxwin.ico'))

        self._createControls()

        # Minimum client size of the main window
        self.SetMinClientSize(wx.Size(400, 300))

        s = SettingsManager()
        mode, pos = s.MainWndStartupPos
        sz = s.MainWndStartupSize
        if mode == wx.ALIGN_NOT:
            if sz == wx.DefaultSize:
                if pos == wx.DefaultPosition:
                    self.Maximize()
                else:
                    self.CenterOnScreen()
            else:
                self.Move(pos)
                self.SetSize(sz)
        else:
            iWScr, iHScr = wx.GetDisplaySize()
            if sz == wx.DefaultSize:
                sz = self.GetSize()
            pt = wx.DefaultPosition
            if (mode & wx.LEFT) == wx.LEFT:
                pt.x = 0
            elif (mode & wx.RIGHT) == wx.RIGHT:
                pt.x = iWScr - sz.GetWidth()
            else:
                pt.x = (iWScr - sz.GetWidth()) / 2
            if (mode & wx.TOP) == wx.TOP:
                pt.y = 0
            elif (mode & wx.BOTTOM) == wx.BOTTOM:
                pt.y = iHScr - sz.GetHeight()
            else:
                pt.y = (iHScr - sz.GetHeight()) / 2
            self.Move(pt)
            self.SetSize(sz)

        self._connectControls()

    def _createControls(self):
        # A Statusbar in the bottom of the window
        self.CreateStatusBar(1)
        sMsg = 'wxPython ' + wx.version() + ' - ' + wx.GetLibraryVersionInfo().VersionString
        self.SetStatusText(sMsg)

        # Add a panel to the frame (needed under Windows to have a nice background)
        pnl = wx.Panel(self, wx.ID_ANY)

        szrMain = wx.BoxSizer(wx.VERTICAL)
        self._btnPrefs = wx.Button(pnl, wx.ID_ANY, 'Edit Settings')
        szrMain.Add(self._btnPrefs, 0, wx.ALL, 5)
        pnl.SetSizer(szrMain)

    def _connectControls(self):
        self.Bind(wx.EVT_SIZE, self._onWindowResized)
        self.Bind(wx.EVT_MOVE, self._onWindowMoved)
        self._btnPrefs.Bind(wx.EVT_BUTTON, self._onBtnEditPrefstClicked)

    def _onWindowResized(self, evt):
        if not self.IsShown():
            return
        s = SettingsManager()
        if self.IsMaximized():
            s.MainWndStartupPos = wx.DefaultPosition
            s.MainWndStartupSize = wx.DefaultSize
        else:
            s.MainWndStartupRect = self.GetRect()

        evt.Skip()

    def _onWindowMoved(self, evt):
        if not self.IsShown():
            return
        SettingsManager().MainWndStartupRect = self.GetRect()

        evt.Skip()

    def _onBtnEditPrefstClicked(self, evt):
        dlg = DlgSettings(self)
        dlg.ShowModal()
        dlg.Destroy()
